src/overlays/actors/ovl_Object_Kankyo/z_object_kankyo.c

```diff
@@ -33,42 +38,689 @@ const ActorInit Object_Kankyo_InitVars = {
     (ActorFunc)ObjectKankyo_Draw,
 };
 
-#endif
+static u16 D_808DE340 = 0;
+
+void ObjectKankyo_SetupAction(ObjectKankyo* this, ObjectKankyoActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+void func_808DBE8C(ObjectKankyo* this) {
+    ObjectKankyo_SetupAction(this, func_808DC18C);
+}
+
+void func_808DBEB0(ObjectKankyo* this, GlobalContext* globalCtx) {
+    s32 i;
+
+    D_808DE5B0 = 0.0f;
+    this->unk_144 = Rand_ZeroOne() * 360.0f;
+    this->unk_148 = Rand_ZeroOne() * 360.0f;
+    if (globalCtx->envCtx.unk_F2[2] == 128) {
+        D_808DE5B0 = 1.0f;
+        this->unk_114E = 1;
+
+        for (i = 0; i < globalCtx->envCtx.unk_F2[2]; i++) {
+            this->unk_14C[i].unk_10 = Rand_ZeroOne() * -200.0f;
+        }
+    } else {
+        this->unk_114E = 0;
+    }
+    ObjectKankyo_SetupAction(this, func_808DCB7C);
+}
+
+void func_808DBFB0(ObjectKankyo* this, GlobalContext* globalCtx) {
+    D_808DE5B0 = 0.0f;
+    this->unk_114E = 0;
+    this->unk_144 = Rand_ZeroOne() * 360.0f;
+    this->unk_148 = Rand_ZeroOne() * 360.0f;
+    this->unk_114C = D_808DE340;
+    D_808DE340++;
+    ObjectKankyo_SetupAction(this, func_808DCBF8);
+}
+
+void func_808DC038(ObjectKankyo* this, GlobalContext* globalCtx) {
+    s16 i;
+
+    this->unk_144 = Rand_ZeroOne() * 360.0f;
+    this->unk_148 = Rand_ZeroOne() * 360.0f;
+    this->unk_114C = 0;
+
+    for (i = 0; i < ARRAY_COUNT(this->unk_14C); i++) {
+        this->unk_14C[i].unk_1C = 0;
+    }
+
+    ObjectKankyo_SetupAction(this, func_808DCDB4);
+}
+
+void ObjectKankyo_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ObjectKankyo* this = THIS;
+    s16 i;
+
+    for (i = 0; i < ARRAY_COUNT(this->unk_14C); i++) {
+        this->unk_14C[i].unk_1C = 0;
+    }
+
+    this->actor.room = -1;
+    switch (this->actor.params) {
+        case 0:
+            func_808DBE8C(this);
+            break;
+
+        case 2:
+            globalCtx->envCtx.unk_F2[2] = 0x80;
+            func_808DBFB0(this, globalCtx);
+            break;
+
+        case 1:
+        case 3:
+            func_808DBEB0(this, globalCtx);
+            break;
+
+        case 4:
+            func_808DC038(this, globalCtx);
+            break;
+    }
+}
+
+void ObjectKankyo_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    ObjectKankyo* this = THIS;
+
+    Actor_MarkForDeath(&this->actor);
+}
+
+void func_808DC18C(ObjectKankyo* this, GlobalContext* globalCtx) {
+    s32 pad;
+    s32 pad2;
+    s32 pad3;
+    f32 sqrt;
```

Comment:
This might shadow a builtin, so I'd rather it were renamed (`root` would do, although @engineer124 can probably advise on what this actually is camera-wise).

Reply:
All this is calculating the magnitude of a vector right? Which is then used to normalize the vector. So `magnitude` as a temp is perfectly fine until proper documentation occurs.
```
magnitude = sqrtf(SQ(x) + SQ(y) + SQ(z));
```

 Also, the reuse of certain temps are very strange in this function (specifically x), so this is a fake match that will need to be fixed later on.